cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -68,8 +73,37 @@ func PerformPostUpgradeTasks(client clientset.Interface, cfg *kubeadmapi.MasterC
 	if err := dns.EnsureDNSAddon(cfg, client); err != nil {
 		errs = append(errs, err)
 	}
+
+	if err := removeDNSDeployment(cfg, client); err != nil {
+		errs = append(errs, err)
+	}
+
 	if err := proxy.EnsureProxyAddon(cfg, client); err != nil {
 		errs = append(errs, err)
 	}
 	return errors.NewAggregate(errs)
 }
+
+func removeDNSDeployment(cfg *kubeadmapi.MasterConfiguration, client clientset.Interface) error {
+	if features.Enabled(cfg.FeatureGates, features.CoreDNS) {
+		ch := make(chan bool, 1)
```

Comment:
I would use `app/util/apiclient.TryRunCommand` for this. It handles the backoff for you automatically